cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
+        if (ad2 == null) {
+            throw new PowsyblException("Unexpected reference to Substation or voltageLevel " + id2
+                + ". It has not been defined in CGMES substations / voltageLevels.");
+        }
+
+        ad1.add(id2);
+        ad2.add(id1);
+    }
+
+    private void buildVoltageLevel(Map<String, List<String>> voltageLevelAdjacency) {
+        Set<String> visitedVoltageLevels = new HashSet<>();
+        voltageLevelAdjacency.keySet().forEach(vl -> {
+            if (!visitedVoltageLevels.contains(vl)) {
+                ArrayList<String> vlAds = allConnected(voltageLevelAdjacency, visitedVoltageLevels, vl);
+                String selectedVoltageLevelId = representativeVoltageLevelId(vlAds);
+                for (String voltageLevelId : vlAds) {
+                    if (!voltageLevelId.equals(selectedVoltageLevelId)) {
+                        voltageLevelMapping.put(voltageLevelId, selectedVoltageLevelId);
                     }
-                });
-        if (!mapping.isEmpty()) {
+                }
+            }
+        });
+        if (!voltageLevelMapping.isEmpty()) {
+            LOG.warn("VoltageLevel id mapping needed for {} voltageLevels: {}",
+                voltageLevelMapping.size(), voltageLevelMapping);
+        }
+    }
+
+    private void buildSubstation(Map<String, List<String>> substationAdjacency) {
+        Set<String> visitedSubstations = new HashSet<>();
+        substationAdjacency.keySet().forEach(sub -> {
+            if (!visitedSubstations.contains(sub)) {
+                ArrayList<String> subAds = allConnected(substationAdjacency, visitedSubstations, sub);
+
+                String selectedSubstationId = representativeSubstationId(subAds);
+                for (String substationId : subAds) {
+                    if (!substationId.equals(selectedSubstationId)) {
+                        substationMapping.put(substationId, selectedSubstationId);
+                    }
+                }
+            }
+        });
+        if (!substationMapping.isEmpty()) {
             LOG.warn("Substation id mapping needed for {} substations: {}",
-                    mapping.size(), mapping);
+                    substationMapping.size(), substationMapping);
         }
     }
 
+    // Given an id (substation / voltageLevel) returns all connected ids (substations / voltageLevels)
+    // Two ids are connected if they are adjacent in the adjacency Map
+    private static ArrayList<String> allConnected(Map<String, List<String>> adjacency, Set<String> visited, String id) {
+        ArrayList<String> allConnected = new ArrayList<>();
+
+        // Insert id in the allConnected list and record it as visited
+        allConnected.add(id);
+        visited.add(id);
+
+        // Expand, adding in each step all non-visited adjacent ids"
+        int k = 0;
+        while (k < allConnected.size()) {
+            String vl0 = allConnected.get(k);
+            if (adjacency.containsKey(vl0)) {
```

Comment:
See my previous comment: using a set instead of a list will make this algorithm more efficient

Reply:
To expand, we need to process the `List` by index as we are increasing it, so we prefer a `Lis`t rather than a `Set.`